core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java

```diff
@@ -0,0 +1,578 @@
+            ManifestFiles.write(format, spec, outputFile, manifestFile.snapshotId());
+        ManifestReader<DataFile> reader =
+            ManifestFiles.read(manifestFile, io, specsById).select(Arrays.asList("*"))) {
+      return StreamSupport.stream(reader.entries().spliterator(), false)
+          .map(entry -> writeDataFileEntry(entry, spec, sourcePrefix, targetPrefix, writer))
+          .collect(Collectors.toList());
+    }
+  }
+
+  /**
+   * Rewrite a delete manifest, replacing path references.
+   *
+   * @param manifestFile source delete manifest to rewrite
+   * @param outputFile output file to rewrite manifest file to
+   * @param io file io
+   * @param format format of the manifest file
+   * @param specsById map of partition specs by id
+   * @param sourcePrefix source prefix that will be replaced
+   * @param targetPrefix target prefix that will replace it
+   * @param stagingLocation staging location for rewritten files (referred delete file will be
+   *     rewritten here)
+   * @return a copy plan of content files in the manifest that was rewritten
+   */
+  public static RewriteResult<DeleteFile> rewriteDeleteManifest(
+      ManifestFile manifestFile,
+      OutputFile outputFile,
+      FileIO io,
+      int format,
+      Map<Integer, PartitionSpec> specsById,
+      String sourcePrefix,
+      String targetPrefix,
+      String stagingLocation)
+      throws IOException {
+    PartitionSpec spec = specsById.get(manifestFile.partitionSpecId());
+    try (ManifestWriter<DeleteFile> writer =
+            ManifestFiles.writeDeleteManifest(format, spec, outputFile, manifestFile.snapshotId());
+        ManifestReader<DeleteFile> reader =
+            ManifestFiles.readDeleteManifest(manifestFile, io, specsById)
+                .select(Arrays.asList("*"))) {
+      return StreamSupport.stream(reader.entries().spliterator(), false)
+          .map(
+              entry ->
+                  writeDeleteFileEntry(
+                      entry, spec, sourcePrefix, targetPrefix, stagingLocation, writer))
+          .reduce(new RewriteResult<>(), RewriteResult::append);
+    }
+  }
+
+  private static Pair<String, String> writeDataFileEntry(
+      ManifestEntry<DataFile> entry,
+      PartitionSpec spec,
+      String sourcePrefix,
+      String targetPrefix,
+      ManifestWriter<DataFile> writer) {
+    DataFile dataFile = entry.file();
+    String sourceDataFilePath = dataFile.location();
+    Preconditions.checkArgument(
+        sourceDataFilePath.startsWith(sourcePrefix),
+        "Encountered data file %s not under the source prefix %s",
+        sourceDataFilePath,
+        sourcePrefix);
+    String targetDataFilePath = newPath(sourceDataFilePath, sourcePrefix, targetPrefix);
+    DataFile newDataFile =
+        DataFiles.builder(spec).copy(entry.file()).withPath(targetDataFilePath).build();
+    appendEntryWithFile(entry, writer, newDataFile);
+    return Pair.of(sourceDataFilePath, newDataFile.location());
```

Comment:
nit: Looks like writeDataFileEntry and writeDeleteFIle entry returns Pair and RewriteResult respectively, same for all the caller on the stack, I am wondering if we shall use the generic RewriteResult for both?